core/sr-primitives/src/offence.rs

```diff
@@ -0,0 +1,128 @@
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Common traits and types that are useful for describing offences for usage in environments
+//! that use staking.
+
+use rstd::vec::Vec;
+
+use codec::{Encode, Decode};
+use crate::Perbill;
+
+/// The kind of an offence, is a byte string representing some kind identifier
+/// e.g. `b"im-online:offlin"`, `b"babe:equivocatio"`
+// TODO [slashing]: Is there something better we can have here that is more natural but still
+// flexible? as you see in examples, they get cut off with long names.
+pub type Kind = [u8; 16];
+
+/// Number of times the offence of this authority was already reported in the past.
+///
+/// Note we don't buffer offence reporting so everytime we see a new offence
+/// of the same kind, we will report past authorities again.
+/// This counter keeps track of how many times the authority was already reported in the past,
+/// so that we can slash it accordingly.
+pub type OffenceCount = u32;
+
+/// A type that represents a point in time on an abstract timescale.
+///
+/// This type is not tied to a particular timescale and can be used for representing instants on
+/// different timescales. Examples of such timescales would be the following: for GRANDPA is a round
+/// ID, for BABE it's a slot number and for offline it's a session index. The only requirement is
+/// that such timescale could be represented by a single `u128` value.
+pub type TimeSlot = u128;
+
+/// A trait implemented by an offence report.
+///
+/// This trait assumes that the offence is legitimate and was validated already.
+///
+/// Examples of offences include: a BABE equivocation or a GRANDPA unjustified vote.
+pub trait Offence<Offender> {
+	/// Identifier which is unique for this kind of an offence.
+	const ID: Kind;
+
+	/// The list of all offenders involved in this incident.
+	///
+	/// The list has no duplicates, so it is rather a set.
+	fn offenders(&self) -> Vec<Offender>;
+
+	/// The session index which is used for querying the validator set for the `slash_fraction`
+	/// function.
+	fn current_era_start_session_index(&self) -> u32; // TODO [slashing]: Should be a SessionIndex.
+
+	/// Return a validators count at the time when the offence took place.
+	fn validators_count(&self) -> u32;
+
+	/// A point in time when this offence happened.
```

Comment:
how are offenses which happened long ago but were only caught recently handled?

Reply:
If the node can produce the proof we will accept it, and afaik we will only produce proofs for not-pruned states initially (i.e. we will call into the runtime at specific block and generate the proof).
So we don't plan to store tries in any kind of DB initially.

Reply:
what I mean is about the point in time.

We may have a bonding period of 3 months but a rolling window is probably not going to be that long.

What happens when we catch an offense that happened almost 3 months ago?